File: riveredge-backend/src/apps/kuaioa/constants/general_signoff_business_types.py
"""通用会签 / 自定义审批业务类型（WP-04B）。

通过 form template / request 的 business_type 绑定字段 schema 与审批声明；
不为每类表单复制 CRUD。让步放行继续走既有 collaboration.concession，不在此重复登记。
"""

from __future__ import annotations

from typing import FrozenSet

# code → 中文名称（产品通用，禁止客户公司名）
GENERAL_SIGNOFF_BUSINESS_TYPES: dict[str, str] = {
    "five_m_change": "5M 变更报告",
    "material_request": "物料申请",
    "sample_inspection": "样品检验",
    "tech_work_contact": "技术工作联系",
    "confirmation": "确认书",
    "review_sheet": "评审单",
    "material_issue": "领料单",
}

GENERAL_SIGNOFF_BUSINESS_TYPE_CODES: FrozenSet[str] = frozenset(
    GENERAL_SIGNOFF_BUSINESS_TYPES.keys()
)


def is_valid_general_signoff_business_type(code: str | None) -> bool:
    raw = (code or "").strip()
    if not raw:
        return True  # 空 = 普通自定义申请
    return raw in GENERAL_SIGNOFF_BUSINESS_TYPE_CODES


def list_general_signoff_business_types() -> list[dict[str, str]]:
    return [
        {"code": code, "name": name}
        for code, name in GENERAL_SIGNOFF_BUSINESS_TYPES.items()
    ]

File: riveredge-frontend/src/apps/kuaioa/constants/generalSignoffBusinessTypes.ts
/**
 * 轻办公通用会签业务类型（WP-04B）
 * 与后端 general_signoff_business_types 对齐；让步放行仍走 collaboration/concession。
 */

export const GENERAL_SIGNOFF_BUSINESS_TYPES = [
  { code: 'five_m_change', labelKey: 'app.kuaioa.formBusinessType.five_m_change' },
  { code: 'material_request', labelKey: 'app.kuaioa.formBusinessType.material_request' },
  { code: 'sample_inspection', labelKey: 'app.kuaioa.formBusinessType.sample_inspection' },
  { code: 'tech_work_contact', labelKey: 'app.kuaioa.formBusinessType.tech_work_contact' },
  { code: 'confirmation', labelKey: 'app.kuaioa.formBusinessType.confirmation' },
  { code: 'review_sheet', labelKey: 'app.kuaioa.formBusinessType.review_sheet' },
  { code: 'material_issue', labelKey: 'app.kuaioa.formBusinessType.material_issue' },
] as const;

export type GeneralSignoffBusinessTypeCode =
  (typeof GENERAL_SIGNOFF_BUSINESS_TYPES)[number]['code'];
